Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -0,0 +1,969 @@
+            Contract.Assume(unwrapCandidates.Count == 2, "Should be exactly two candidate Unwrap methods.");
+
+            // We need to find appropriate Unwrap method based on CheckMethod argument type.
+            var firstMethod = (Method)unwrapCandidates[0];
+            var secondMethod = (Method)unwrapCandidates[1];
+
+            Contract.Assume(firstMethod != null && secondMethod != null);
+
+            var genericUnwrapCandidate = firstMethod.IsGeneric ? firstMethod : secondMethod;
+            var nonGenericUnwrapCandidate = firstMethod.IsGeneric ? secondMethod : firstMethod;
+
+            if (checkMethodTaskType.IsGeneric)
+            {
+                // We need to "instantiate" generic first.
+                // I.e. for Task<int> we need to have Unwrap(Task<Task<int>>): Task<int>
+                return genericUnwrapCandidate.GetTemplateInstance(null, checkMethodTaskType.TemplateArguments[0]);
+            }
+
+            return nonGenericUnwrapCandidate;
+        }
+
+        /// <summary>
+        /// Factory method that creates bool CheckException(Exception e)
+        /// </summary>
+        [Pure]
+        private Method CreateCheckExceptionMethod()
+        {
+            Contract.Ensures(Contract.Result<Method>() != null);
+
+            var exnParameter = new Parameter(Identifier.For("e"), SystemTypes.Exception);
+            var checkExceptionMethod =
+                new Method(
+                    declaringType: this.closureClass,
+                    attributes: null,
+                    name: CheckExceptionMethodId,
+                    parameters: new ParameterList(exnParameter),
+                    returnType: SystemTypes.Boolean,
+                    body: new Block(new StatementList()));
+
+            checkExceptionMethod.Body.HasLocals = true;
+
+            checkExceptionMethod.CallingConvention = CallingConventionFlags.HasThis;
+            checkExceptionMethod.Flags |= MethodFlags.Public;
+
+            if (checkExceptionMethod.ExceptionHandlers == null)
+                checkExceptionMethod.ExceptionHandlers = new ExceptionHandlerList();
+
+            return checkExceptionMethod;
+        }
+
+        private void EmitCheckExceptionBody(Method checkExceptionMethod, List<EnsuresExceptional> exceptionalPostconditions)
+        {
+            Contract.Requires(checkExceptionMethod != null);
+
+            Contract.Requires(exceptionalPostconditions != null);
+            Contract.Requires(exceptionalPostconditions.Count > 0);
+
+            // We emit the following method:
+            //   bool CheckException(Exception e) {
+            //     var ex = e as C1;
+            //     if (ex != null) {
+            //       EnsuresOnThrow(predicate)
+            //     }
+            //     else {
+            //       var ex2 = e as AggregateException;
+            //       if (ex2 != null) {
+            //         ex2.Handle(CheckException);
```

Comment:
:exclamation: This seems incorrect. Async/await only unwrap exactly the first exception, and only unwrap one level. I'm willing to create a separate issue to track this.